### PR TITLE
dev/core#109: Add case type category column

### PR DESCRIPTION
**Pull request by reneolivo**
Overview
----------------------------------------
This PR adds a new column category for case types. The reasons is to be able to make case types more extensible for the configuration and management of case types in other extensions.

By default the case types categories can be Workflow (the default), or Vacancy (for application case types, etc).

Before
----------------------------------------

### Case Type list
before-2

### Case Type form
before-3

After
----------------------------------------

### Case Type list
after-2

### Case Type form
after-3

Technical Details
----------------------------------------
### Adding the Vacancy and Workflow categories

A new upgrader was added:

```php
public function upgrade_5_2_beta1($rev) {
  $this->addTask(ts('Upgrade DB to %1: SQL', array(1 => $rev)), 'runSql', $rev);

  $this->addCategoryColumnToCaseType();
  $this->createCaseTypeCategories();
  $this->setDefaultCategoriesForCaseTypes();
}

private function addCategoryColumnToCaseType() {
  // could not add this to the 5.2.beta1.mysql.tpl file because it seems the result is not ready by the
  // time the PHP script runs:
  CRM_Core_DAO::executeQuery('ALTER TABLE civicrm_case_type
    ADD COLUMN category VARCHAR(512)');
}

private function createCaseTypeCategories() {
  // creates the case type category group:
  CRM_Core_BAO_OptionGroup::ensureOptionGroupExists([
    'name' => 'case_type_category',
    'title' => ts('Case Type Category'),
    'is_reserved' => 1,
  ]);
  
  // flush the pseudo constant cache so it includes the newly created option group:
  CRM_Core_PseudoConstant::flush();

  $options = [
    ['name' => 'WORKFLOW', 'label' => ts('Workflow')],
    ['name' => 'VACANCY', 'label' => ts('Vacancy')]
  ];

  foreach ($options as $option) {
    // creates the workflow and vacancy option values:
    CRM_Core_BAO_OptionValue::ensureOptionValueExists([
      'option_group_id' => 'case_type_category',
      'name' => $option['name'],
      'label' => $option['label'],
      'is_active' => TRUE,
      'is_reserved'=> TRUE
    ]);
  }
}

private function setDefaultCategoriesForCaseTypes() {
  $caseTypes = civicrm_api3('CaseType', 'get', [
    'options' => [ 'limit' => 0 ]
  ]);

  foreach ($caseTypes['values'] as $caseType) {
    $category = $caseType['name'] === 'Application' ? 'VACANCY' : 'WORKFLOW';

    // updates all current case types. By default they have the "Workflow" category, except for
    // Application case types, which has the "Vacancy" category:
    civicrm_api3('CaseType', 'create', [
      'id' => $caseType['id'],
      'category' => $category
    ]);
  }
}
```

### Updating the Case Type Schema and DAO

The Case Type Schema was updated so it includes the new category field and references the `case_type_category` pseudo constant (option value).

```xml
<?xml version="1.0" encoding="utf-8" ?>
<table>
  <name>civicrm_case_type</name>
  <!-- ... -->
  <field>
    <name>category</name>
    <title>Category</title>
    <type>int unsigned</type>
    <required>true</required>
    <add>5</add>
    <comment>FK to a civicrm_option_value (case_type_category)</comment>
    <pseudoconstant>
      <optionGroupName>case_type_category</optionGroupName>
    </pseudoconstant>
    <html>
      <type>Select</type>
    </html>
  </field>
  <!-- ... -->
</table>
```

The DAO file was updated by running `php GenCode.php` at the `xml` folder.

### Displaying the category at the case type administration screen

The list of case type categories is provided to the `CaseTypeListCtrl`:

```js
$routeProvider.when('/caseType', {
  templateUrl: '~/crmCaseType/list.html',
  controller: 'CaseTypeListCtrl',
  resolve: {
    // ..
    caseTypeCategories: function (crmApi) {
      return crmApi('OptionValue', 'get', {options: {limit: 0}})
    }
  }
});
```

The result is indexed by value:

```js
crmCaseType.controller('CaseTypeListCtrl', function($scope, crmApi, caseTypes, caseTypeCategories) {
  // ...
  $scope.caseTypeCategoriesIndex = _.indexBy(caseTypeCategories.values, 'value');
  // ...
});
```

And the category label is displayed on the view:

```html
<td>{{caseTypeCategoriesIndex[caseType.category].label}}</td>
```

### New case types

For new case types the Workflow category is automatically selected:

```js
crmCaseType.controller('CaseTypeCtrl', function($scope, crmApi, apiCalls) {
  // ...
  isNewCaseType = !apiCalls.caseType,
  // ...

  if (isNewCaseType) {
    $scope.caseType = _.cloneDeep(newCaseTypeTemplate);
    $scope.caseType.category = defaultCategory.value;
  } else {
    $scope.caseType = apiCalls.caseType;
  }

  // ...
}
```

------

Gitlab issue: https://lab.civicrm.org/dev/core/issues/109

### CODE REVIEW

**Comment by tunbola** [comment]
@reneolivo, I have just few notes:
Doesn't this PR need to reference a Core JIRA Issue or something?
The functions used by the upgrader are written as protected in the PR details but they are actually private in the code.

File: ang/crmCaseType.js
Comment (igorpavlov) [inline ang/crmCaseType.js]: Would we like to sort the categories?

File: ang/crmCaseType.js
Comment (igorpavlov) [inline ang/crmCaseType.js]: There was an explicit change that checked if this API call exists here 9625aad#diff-1e51710f6a7a59a5c0c0db9a53d71b89R139. Is there a reason why we don't want to check it anymore?

File: tests/karma/unit/crmCaseTypeSpec.js
Comment (igorpavlov) [inline tests/karma/unit/crmCaseTypeSpec.js]: Would this variable name be more self explanatory? `expectedDefaultCategory`

File: tests/karma/unit/crmCaseTypeSpec.js
Comment (igorpavlov) [inline tests/karma/unit/crmCaseTypeSpec.js]: Would such a variable name be more self explanatory? `expectedCaseTypeCategoriesIndexed`

File: tests/karma/unit/crmCaseTypeSpec.js
Comment (igorpavlov) [inline tests/karma/unit/crmCaseTypeSpec.js]: `API` is an abbreviation, could you please capitalise it?

File: tests/karma/unit/crmCaseTypeSpec.js
Comment (igorpavlov) [inline tests/karma/unit/crmCaseTypeSpec.js]: Should one of case type categories be `"is_default": "1"`?

**Review by igorpavlov-zz — CHANGES_REQUESTED** [review]

**Comment by igorpavlov** [comment]
@reneolivo Reviewed a left a couple of questions and comments to JS and HTML files.

File: ang/crmCaseType.js
Comment (reneolivo) [reply ang/crmCaseType.js]: The check is still being done:
https://github.com/civicrm/civicrm-core/pull/12098/files#diff-f42508faf29afc0f8ecd02bcb3d6be66R138

File: tests/karma/unit/crmCaseTypeSpec.js
Comment (reneolivo) [reply tests/karma/unit/crmCaseTypeSpec.js]: Good question. Test-wise I don't think it's necessary. The question is if we need to make an actual option value to be the default one. Will investigate.

**Comment by reneolivo** [comment]
@tunbola :
> Doesn't this PR need to reference a Core JIRA Issue or something?

I just registered to the civicrm's gitlab platform, but I'm waiting for admin approval. As soon as I have the approval I'll create the ticket and add the ticket number to the commits.

> The functions used by the upgrader are written as protected in the PR details but they are actually private in the code.

Thanks for spotting that. Fixed it.

> Also looking at https://docs.civicrm.org/user/en/latest/case-management/set-up/, This change might need some update to the user guide documentation as regards creating a case and adding case type categories.

I agree. This change should be documented. I'll wait until the gitlab ticket gets approved to start working on the docs so the PR to the docs will focus on the latest changes done here.

**Comment by eileenmcnaughton** [comment]
I've moved this to being against master since we would not add a feature to the 5.2 rc (or any rc). The upgrade script will need to be updated once it is clear which release this will go into

It's not clear at the moment that this is approved in concept for core - so it would make sense to open a gitlab issue & close this & re-open when that is clear (having said that I think it's OK for PRs to be open for a short period for actual discussion even when the approval is not clear - but if this is not clarified in the next week or so we should track in gitlab rather than the PR queue)

**Comment by eileenmcnaughton** [comment]
As an aside - looking at this I'm wondering if part of our proposal process is that any proposal that involves a schema change should require an email to the dev list

**Comment by colemanw** [comment]
Please see my comments on the [GitLab issue](https://lab.civicrm.org/dev/core/issues/109). If this is needed for hooks, then I think a custom field ought to suffice here.

**Comment by jamienovick** [comment]
@reneolivo @colemanw @eileenmcnaughton added comments on gitlab...

**Comment by jamienovick** [comment]
I don't think Vacancy + Workflow should be defaults. I think perhaps something more generic such as:

- Internal workflow 
- Service delivery

or 

- Internal
- External

if we did go ahead with this.

**Comment by eileenmcnaughton** [comment]
I'm going to close this out for now - if /when the concept is approved it should be re-opened.

It's not really reviewable at the moment as that question has not been resolved.

**Pull request closed without merging**
